worker/peergrouper/machinetracker.go

```diff
@@ -84,21 +83,40 @@ func (m *machineTracker) Addresses() []network.Address {
 	return out
 }
 
-// SelectMongoHostPort returns the best hostport for the machine for
-// MongoDB use, perhaps using the space provided.
-func (m *machineTracker) SelectMongoHostPort(port int, space network.SpaceName) string {
+// SelectMongoAddress returns the best address on the machine for
+// MongoDB peer use, using the input space if non-empty.
+// If no space is provided, and there is more than one non-machine-local
+// address on the machine, then an error is returned.
+// We require an HA space configuration to be set in this case.
+func (m *machineTracker) SelectMongoAddress(port int, space network.SpaceName) (string, error) {
 	m.mu.Lock()
 	hostPorts := network.AddressesWithPort(m.addresses, port)
 	m.mu.Unlock()
 
 	if space != "" {
-		addr := mongo.SelectPeerHostPortBySpace(hostPorts, space)
-		logger.Debugf("machine %q selected address %q by space %q from %v", m.id, addr, space, hostPorts)
-		return addr
+		addrs, ok := network.SelectHostPortsBySpaceNames(hostPorts, space)
+		if ok {
+			addr := addrs[0].NetAddr()
+			logger.Debugf("machine %q selected address %q by space %q from %v", m.id, addr, space, hostPorts)
+			return addr, nil
+		}
+		// If we end up here, then there are no addresses available in the
+		// specified space. This should not happen, because the configured
+		// space is used as a constraint when first enabling HA.
+		logger.Debugf("Failed to select hostPort by space - trying by scope from %+v", hostPorts)
```

Comment:
I think if a space is configured, and we can't get an address in it, we should *not* fall back to a cloud-local address. What do you think? Do you think it is better to at least try *some* address?
It does feel like it merits a Warning rather than a Debug.

Reply:
To my mind, the only way this path could be traversed is if (as we are planning to allow) the HA space config is updated post-bootstrap. As per my comment, it constrains provisioning, so we couldn't get here via issuing _enable-ha_.

So then +not+ falling back and throwing out means that you do change the HA space config value, but if unsatisfied then nothing about the peer group changes. I think I need to defer to your judgement here. It becomes a choice between:
- There is HA space config that we have taken our best shot at satisfying or;
- There is HA space config and the peer group represents a prior config until things are fixed.

Will change the level to Warning.

Reply:
I would like "juju controller-config ha-space=BLAH" to fail if it doesn't think all current controller machines are in space BLAH, which should also narrow the chance for this to trigger.
I'm hesitant with "fallback to something that is maybe ok". The issue is that it is probably true, that the set of circumstances where things will "work" is slightly greater if we have the fallback. The concern is that these are "working" by accident, so people will think it is working when it may not actually be. (We found an address, which we then advertise, but the firewall rules actually don't let traffic through.)
And then discovering that there is a problem has to be someone noticing that Mongo is failing to join the replicaset, and digging through logs to see that we fell back to an address that wasn't requested.